File: CS/NavBarDataBindingHelper.cs
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Windows;
using DevExpress.Xpf.NavBar;

namespace NavBar_DataBinding {

    public class NavBarDataBindingHelper : DependencyObject {
        public static readonly DependencyProperty ItemsSourceProperty;

        static NavBarDataBindingHelper() {
            ItemsSourceProperty = DependencyProperty.RegisterAttached("ItemsSource", 
                typeof(IEnumerable), typeof(NavBarDataBindingHelper), 
                new FrameworkPropertyMetadata(null, OnItemsSourceChanged));
        }

        public static void SetItemsSource(DependencyObject element, IEnumerable<NavBarGroup> value) {
            element.SetValue(NavBarDataBindingHelper.ItemsSourceProperty, value);
        }

        public static IEnumerable GetItemsSource(DependencyObject element) {
            return (IEnumerable)element.GetValue(NavBarDataBindingHelper.ItemsSourceProperty);
        }

        static Dictionary<NavBarControl, Binder> navBars = new Dictionary<NavBarControl, Binder>();

        static void OnItemsSourceChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            NavBarControl navBar = d as NavBarControl;
            if (navBar == null)
                return;
            if (navBars.ContainsKey(navBar)) {
                navBars[navBar].UnsubscribeEventsIfNeeded();
                navBars.Remove(navBar);
            }
            if (e.NewValue != null)
                navBars[navBar] = new Binder(navBar, NavBarDataBindingHelper.GetItemsSource(navBar));
        }
    }

    class Binder {
        NavBarControl navBar;
        INotifyCollectionChanged collection;

        public Binder(NavBarControl navBar, IEnumerable source) {
            this.navBar = navBar;

            navBar.Groups.Clear();
            foreach (NavBarGroup group in source)
                navBar.Groups.Add(group);

            collection = source as INotifyCollectionChanged;
            if (collection != null)
                collection.CollectionChanged += 
                    new NotifyCollectionChangedEventHandler(BindingHelper_CollectionChanged);
        }

        public void UnsubscribeEventsIfNeeded() {
            if (collection != null)
                collection.CollectionChanged -= 
                    new NotifyCollectionChangedEventHandler(BindingHelper_CollectionChanged);
        }

        void BindingHelper_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e) {
            if (e.NewItems != null)
                foreach (NavBarGroup group in e.NewItems)
                    navBar.Groups.Add(group);
            if (e.OldItems != null)
                foreach (NavBarGroup group in e.OldItems)
                    navBar.Groups.Remove(group);
        }
    }
}

File: VB/NavBarDataBindingHelper.vb
Imports Microsoft.VisualBasic
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.Windows
Imports DevExpress.Wpf.NavBar

Namespace NavBar_DataBinding

	Public Class NavBarDataBindingHelper
		Inherits DependencyObject
		Public Shared ReadOnly ItemsSourceProperty As DependencyProperty

		Shared Sub New()
			ItemsSourceProperty = DependencyProperty.RegisterAttached("ItemsSource", GetType(IEnumerable), GetType(NavBarDataBindingHelper), New FrameworkPropertyMetadata(Nothing, AddressOf OnItemsSourceChanged))
		End Sub

		Public Shared Sub SetItemsSource(ByVal element As DependencyObject, ByVal value As IEnumerable(Of NavBarGroup))
			element.SetValue(NavBarDataBindingHelper.ItemsSourceProperty, value)
		End Sub

		Public Shared Function GetItemsSource(ByVal element As DependencyObject) As IEnumerable
			Return CType(element.GetValue(NavBarDataBindingHelper.ItemsSourceProperty), IEnumerable)
		End Function

		Private Shared navBars As New Dictionary(Of NavBarControl, Binder)()

		Private Shared Sub OnItemsSourceChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim navBar As NavBarControl = TryCast(d, NavBarControl)
			If navBar Is Nothing Then
				Return
			End If
			If navBars.ContainsKey(navBar) Then
				navBars(navBar).UnsubscribeEventsIfNeeded()
				navBars.Remove(navBar)
			End If
			If e.NewValue IsNot Nothing Then
				navBars(navBar) = New Binder(navBar, NavBarDataBindingHelper.GetItemsSource(navBar))
			End If
		End Sub
	End Class

	Friend Class Binder
		Private navBar As NavBarControl
		Private collection As INotifyCollectionChanged

		Public Sub New(ByVal navBar As NavBarControl, ByVal source As IEnumerable)
			Me.navBar = navBar

			navBar.Groups.Clear()
			For Each group As NavBarGroup In source
				navBar.Groups.Add(group)
			Next group

			collection = TryCast(source, INotifyCollectionChanged)
			If collection IsNot Nothing Then
				AddHandler collection.CollectionChanged, AddressOf BindingHelper_CollectionChanged
			End If
		End Sub

		Public Sub UnsubscribeEventsIfNeeded()
			If collection IsNot Nothing Then
				RemoveHandler collection.CollectionChanged, AddressOf BindingHelper_CollectionChanged
			End If
		End Sub

		Private Sub BindingHelper_CollectionChanged(ByVal sender As Object, ByVal e As NotifyCollectionChangedEventArgs)
			If e.NewItems IsNot Nothing Then
				For Each group As NavBarGroup In e.NewItems
					navBar.Groups.Add(group)
				Next group
			End If
			If e.OldItems IsNot Nothing Then
				For Each group As NavBarGroup In e.OldItems
					navBar.Groups.Remove(group)
				Next group
			End If
		End Sub
	End Class
End Namespace